internal/controllers/core/cluster/reconciler.go

```diff
@@ -22,20 +37,144 @@ func (r *Reconciler) CreateBuilder(mgr ctrl.Manager) (*builder.Builder, error) {
 	return b, nil
 }
 
-func NewReconciler(ctrlClient ctrlclient.Client) *Reconciler {
+func NewReconciler(ctrlClient ctrlclient.Client, store store.RStore, localDockerEnv docker.LocalEnv) *Reconciler {
 	return &Reconciler{
-		ctrlClient: ctrlClient,
+		ctrlClient:     ctrlClient,
+		store:          store,
+		localDockerEnv: localDockerEnv,
+		connections:    make(map[types.NamespacedName]*connection),
 	}
 }
 
 func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
 	nn := request.NamespacedName
+	ctx = logger.CtxWithLogHandler(ctx, logWriter{store: r.store})
 
 	var obj v1alpha1.Cluster
 	err := r.ctrlClient.Get(ctx, nn, &obj)
 	if err != nil && !apierrors.IsNotFound(err) {
+		delete(r.connections, nn)
+		return ctrl.Result{}, err
+	}
+
+	connection, hasConnection := r.connections[nn]
+	if hasConnection {
+		// If the spec changed, delete the connection and recreate it.
+		if !apicmp.DeepEqual(connection.spec, obj.Spec) {
+			delete(r.connections, nn)
+			connection = nil
+			hasConnection = false
+		}
+	}
+
+	if !hasConnection {
+		// Create the initial connection to the cluster.
+		if obj.Spec.Connection != nil && obj.Spec.Connection.Kubernetes != nil {
+			connection = r.createKubernetesConnection(ctx, obj.Spec.Connection.Kubernetes)
+		} else if obj.Spec.Connection != nil && obj.Spec.Connection.Docker != nil {
+			connection = r.createDockerConnection(ctx, obj.Spec.Connection.Docker)
+		}
+		connection.createdAt = time.Now()
+		connection.spec = obj.Spec
+	}
+
+	// TODO(nick): Populate the cluster architecture.
+	status := connection.toStatus()
+	err = r.maybeUpdateStatus(ctx, &obj, status)
+	if err != nil {
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
+
+// Creates a docker connection from the spec.
+func (r *Reconciler) createDockerConnection(ctx context.Context, obj *v1alpha1.DockerClusterConnection) *connection {
+	// If no Host is specified, use the default Env from environment variables.
+	env := docker.Env(r.localDockerEnv)
+	if obj.Host != "" {
+		env = docker.Env{Host: obj.Host}
```

Comment:
Won't you generally also need TLS client certs in this case?

Reply:
yep! i think there are some notes on this in the API structs - we're eventually going to need to support all the major docker client options. But for now i just wanted to have a placeholder here to demo what it would look like.

Reply:
(my current old-school remote cluster uses TCP+an authenticated tailscale tunnel without any client certs)